Handle failed and abandoned DNS query and/or Socket connection NpgsqlConnector Tasks

baal2000 opened the pull request:
# Technical details

This pull request addresses https://github.com/npgsql/npgsql/issues/4149

- Npgsql's own `TaskExtensions` class wraps a `Task` created by `Socket.ConnectAsync` and/or `Dns.GetHostAddressesAsync` to allow cancellation of a non-cancelable Task or enforce a timeout.
- The original `Task` becomes abandoned in the process and may fail later and trigger `TaskScheduler.UnobservedTaskException` event.   

# Acceptance criteria

- Remove duplicate and unused functionality in `TaskExtensions.cs` and `NpgSqlConnector.cs`.
- Remove existing Task extension methods from TaskExtensions.cs and replace with .NET 6.0 own Task.WaitAsync extension where needed.
- Convert `TaskExtensions` to a non-extension class `TaskTimeoutAndCancellation`. 
- Ensure that the tasks that `TaskTimeoutAndCancellation` class prevents unobserved Task exceptions from leaking as `TaskScheduler.UnobservedTaskException` event.
- Unit test `TaskTimeoutAndCancellation` functionality and `TaskScheduler.UnobservedTaskException` event avoidance logic.

baal2000 commented:
@roji please approve the workflow

baal2000 commented:
@roji please trigger the pipeline run.

baal2000 commented on src/Npgsql/TaskExtensions.cs:
```diff
@@ -8,122 +9,127 @@ namespace Npgsql
-        }
+        
+#if NET6_0_OR_GREATER
+        static Task WithCancellationAndTimeout(Task task, NpgsqlTimeout timeout, CancellationToken cancellationToken)
```
@roji FYI I'll move this line out of if/else/endif in my next commit to avoid duplication
```
        static Task WithCancellationAndTimeout(Task task, NpgsqlTimeout timeout, CancellationToken cancellationToken) =>
#if NET6_0_OR_GREATER
            WithCancellationAndTimeoutImpl(task, timeout, cancellationToken);
#else
            WithCancellationAndTimeoutImplLegacy(task, timeout, cancellationToken);
#endif
```

baal2000 commented:
@roji 
Please review the PR.
The scope of the change is wider than just the bug fix and also includes removal of unused functionality in TaskExtensions so that it is easier to unit test.

If there is anythig that we need "just in case" please let me know.

baal2000 commented:
@roji please trigger the build after the recent update.

baal2000 commented:
@roji @vonzshik  Please review.

baal2000 commented:
@roji 
Sorry to interrupt, but how far is my line in the queue for a review?

roji commented:
@baal2000 sorry for not reacting sooner to this, it's been busy... I promise I'll get around to it in the next 2 days!

roji commented:
Note: this may need to wait for the weekend (/cc @vonzshik)

vonzshik commented:
Definitely no queue here, just a lot of personal stuff to deal with :(

Hopefully, I'll get some free time on the weekend.

vonzshik commented on src/Npgsql/TaskTimeoutAndCancellation.cs:
```diff
@@ -0,0 +1,134 @@
+            {
+                // Prevent unobserved Task notifications by observing the failed Task exception.
+                // To test: comment the next line out and re-run TaskExtensionsTest.DelayedFaultedTaskCancellation.
+                _ = task!.ContinueWith( t => _ = t.Exception, CancellationToken.None, TaskContinuationOptions.OnlyOnFaulted, TaskScheduler.Current);
```
In some cases, `getTaskFunc` can throw an exception, and not a task wrapping an exception (this can happen if the exception is thrown from a method not marked as `async`), which will lead to an nre here. We should make a continuation only after we've made sure the task isn't `null`.

baal2000 replied:
Good catch. Should just take `getTaskFunc` outside of try/catch:
```
            task = getTaskFunc(combinedCts?.Token ?? cancellationToken);
            try
            {
                    ...
            }
            catch ....
```

baal2000 replied:
I have posted the code update.

vonzshik reviewed (changes requested):
All in all, looks OK. One possible NRE to fix and it can be merged.

@roji perhaps you have something to add?

baal2000 commented:
@vonzshik thanks for the review. I rebased/resolved conflicts and also changed the test module to the file-scoped namespace. Please trigger the workflow.

baal2000 commented:
@roji is your review coming next or the PR can be merged?

roji commented on src/Npgsql/Internal/NpgsqlConnector.cs:
```diff
@@ -995,41 +1005,22 @@ async Task ConnectAsync(NpgsqlTimeout timeout, CancellationToken cancellationTok
+            static Task OpenSocketConnectionAsync(Socket socket, EndPoint endpoint, NpgsqlTimeout perIpTimeout, CancellationToken cancellationToken)
+            {
+                // .NET 5.0 added cancellation support to ConnectAsync, which allows us to implement real
+                // cancellation and timeout. On older TFMs, we fake-cancel the operation, i.e. stop waiting
```
See comment above about us always fake-cancelling even on new .NET TFMs

roji commented on src/Npgsql/Internal/NpgsqlConnector.cs:
```diff
@@ -995,41 +1005,22 @@ async Task ConnectAsync(NpgsqlTimeout timeout, CancellationToken cancellationTok
+                // .NET 5.0 added cancellation support to ConnectAsync, which allows us to implement real
+                // cancellation and timeout. On older TFMs, we fake-cancel the operation, i.e. stop waiting
+                // and raise the exception, but the actual connection task is left running.
+                Task ConnectAsync(CancellationToken ct)
```
A local function here seems a bit heavy - it seems simpler to just do something like:

```c#
#if NET5_0_OR_GREATER
                var connectTask = socket.ConnectAsync(endpoint, cancellationToken).AsTask();
#else
                var connectTask = socket.ConnectAsync(endpoint);
#endif
```

baal2000 replied:
The reason for a local function is the same as explained [here](https://github.com/npgsql/npgsql/pull/4173/files#r776901310)

roji replied:
Makes sense, thanks.

roji commented on src/Npgsql/TaskTimeoutAndCancellation.cs:
```diff
@@ -0,0 +1,132 @@
+        => task.WaitAsync(timeout.CheckAndGetTimeLeft(), cancellationToken);
+#endif
+
+    static async Task WithCancellationAndTimeoutImplLegacy(Task task, NpgsqlTimeout timeout, CancellationToken cancellationToken)
```
As written elsewhere, I think this code should be #ifdef'd to pre-.NET 6.0.

The simplest seems to be to copy the implementation of the .NET 6.0 Task.WaitAsync into an extension method in our project. This would allow the code here to simply use WaitAsync without caring whether it's coming from .NET 6.0 or from our own copy. We've done this for various other APIs which were added in later TFMs, see the "shims" directory for some examples (this is where I'd put it).

baal2000 replied:
Let's agree that the class's public interface is agnostic to whether it is .NET 6.0 or an earlier runtime.
It is not an extension class. The only reason that, as explained earlier, this class provides a clear way of canceling the Task that is about to become abandoned. 
We are not re- implementing NET 6.0 Task.WaitAsync extension API. That API is there but is easy to misuse BTW. It carries a risk of creating the same complex problem as we are trying to fix here: abandoning the original Task.

roji replied:
I think in my comment above I confused the higher-level task of preventing the unobserved exception and the lower-level task of implementing the cancellation and timeout - you're right that WaitAsync does the former but not the latter.

But leaving aside the former and concentrating on the factoring of the latter... The current code as it is has TaskTimeoutAndCancellation.WithCancellationAndTimeout, which simply wraps Task.WaitAsync for .NET6.0, and WithCancellationAndTimeout, which provides the identical functionality for older TFMs (and with the parameters etc.). In other words, WithCancellationAndTimeout seems to provide the exact same functionality that .NET 6.0 WaitAsync does - it really is a backwards-compat simple shim. So unless I'm mistaken, there's nothing preventing WithCancellationAndTimeout from being implemented as a Task.WaitAsync extension method, providing the exact same API as the .NET 6.0 but for earlier TFMs, no? This would simplify things and move the legacy code out to the shims directory.

baal2000 replied:
I think I understand. Will make a change and see if that's what you want.

roji replied:
Looks great, thanks! Could you please move the legacy WaitAsync implementation to a TaskExtensions.cs under the Shims directory? That's how we usually take care of this kind of thing.

baal2000 replied:
Will do.

baal2000 replied:
Pushed the update. Please run the checks.

baal2000 replied:
One more thing: do we need the copyright notices for each new module? There weren't any in the former TaskExtensions.cs and I kept it that way.

roji reviewed (changes requested):
Sorry for taking so long to review this - it's been busy period recently.

Overall things look good, though I think factoring-wise some things could be improved - see the comments. I'll be available now for quicker review iterations (be sure to re-request a review once changes are made).

baal2000 commented on src/Npgsql/Internal/NpgsqlConnector.cs:
```diff
@@ -937,13 +937,23 @@ void Connect(NpgsqlTimeout timeout)
-            ? new EndPoint[] { new UnixDomainSocketEndPoint(socketPath) }
-            : (await GetHostAddressesAsync(timeout, cancellationToken))
-            .Select(a => new IPEndPoint(a, Port)).ToArray();
+        async Task ConnectAsync(NpgsqlTimeout timeout, CancellationToken cancellationToken)
```
done

baal2000 commented on src/Npgsql/Internal/NpgsqlConnector.cs:
```diff
@@ -937,13 +937,23 @@ void Connect(NpgsqlTimeout timeout)
+#endif
+
+            // Note that there aren't any timeout-able DNS methods before .NET 6.0. 
+            // .NET 6.0 added cancellation support to GetHostAddressesAsync, which allows us to implement real
```
Done

baal2000 commented on src/Npgsql/Internal/NpgsqlConnector.cs:
```diff
@@ -937,13 +937,23 @@ void Connect(NpgsqlTimeout timeout)
+            // and raise the exception, but the actual connection task is left running.
+            var endpoints = NpgsqlConnectionStringBuilder.IsUnixSocket(Host, Port, out var socketPath)
+                ? new EndPoint[] { new UnixDomainSocketEndPoint(socketPath) }
+                : (await TaskTimeoutAndCancellation.WaitAsync(GetHostAddressesAsync, timeout, cancellationToken))
```
Whether the Task is cancellable or not (framework/platform dependent), this is the only API I can think of that makes the best effort to cancel it with a combined cancellation token passed in (a combination of the original token and a timeout-based token). Do I miss a reason not to do that?

roji replied:
My bad, I confused the concern of applying the timeout via the cancellation token (which requires the lambda as it needs to pass a linked cancellation token) and the other stuff (see https://github.com/npgsql/npgsql/pull/4173#discussion_r776990767).

baal2000 commented on src/Npgsql/Internal/NpgsqlConnector.cs:
```diff
@@ -995,41 +1005,22 @@ async Task ConnectAsync(NpgsqlTimeout timeout, CancellationToken cancellationTok
-            // cancellation and timeout. On older TFMs, we fake-cancel the operation, i.e. stop waiting
-            // and raise the exception, but the actual connection task is left running.
-
+            static Task OpenSocketConnectionAsync(Socket socket, EndPoint endpoint, NpgsqlTimeout perIpTimeout, CancellationToken cancellationToken)
```
done

baal2000 commented on src/Npgsql/Internal/NpgsqlConnector.cs:
```diff
@@ -995,39 +1005,20 @@ async Task ConnectAsync(NpgsqlTimeout timeout, CancellationToken cancellationTok
-                return TaskExtensions.ExecuteWithTimeout(
-                    ct => socket.ConnectAsync(endpoint, ct).AsTask(),
-                    perIpTimeout, cancellationToken);
+                    return socket.ConnectAsync(endpoint, ct).AsTask();
```
done

baal2000 commented on src/Npgsql/TaskTimeoutAndCancellation.cs:
```diff
@@ -0,0 +1,132 @@
+
+        Task WithCancellationAndTimeoutLocal(Task t)
+        {
+            if (useLegacyImplementation)
```
OK

baal2000 commented on src/Npgsql/TaskTimeoutAndCancellation.cs:
```diff
@@ -0,0 +1,132 @@
+    /// <param name="useLegacyImplementation">An optional <see cref="Boolean"/> parameter for testing only. If set to true the method uses pre-dotnet 6.0 implementation for cancellation and timeout enforcement.</param>
+    /// <typeparam name="TResult">The result <see cref="Type"/>.</typeparam>
+    /// <returns>The <see cref="Task{TResult}"/> representing the asynchronous wait.</returns>
+    internal static async Task<TResult> WaitAsync<TResult>(Func<CancellationToken, Task<TResult>> getTaskFunc, NpgsqlTimeout timeout, CancellationToken cancellationToken, bool useLegacyImplementation = false)
```
Maube be a silly question: does the repo runs the test pipelines for all supported .NET platforms (.e. 3.1 ... 6.0)?

baal2000 replied:
Removed `useLegacyImplementation`

roji replied:
@baal2000 it doesn't currently, as we haven't had any issues with older TFMs - but that would be quite easy to do.

baal2000 replied:
I was afraid that is the case and added `useLegacyImplementation` flag because of that. Otherwise I would not be comfortable commiting new code to the repository knowing the legacy TFMs are not regression-tested.

> that would be quite easy to do

Could you make a plan for that?

roji replied:
@baal2000 I fully understand, have opened #4258 to track this - I'll take a look at it.

baal2000 commented on src/Npgsql/TaskTimeoutAndCancellation.cs:
```diff
@@ -0,0 +1,132 @@
+    }
+    
+#if NET6_0_OR_GREATER
+    static Task WithCancellationAndTimeoutImpl(Task task, NpgsqlTimeout timeout, CancellationToken cancellationToken)
```
inlined

baal2000 commented:
@roji 
Thanks for the review. Made most of the recommended changes except for sticking with the non-extension-based helper class API.

roji commented on src/Npgsql/TaskTimeoutAndCancellation.cs:
```diff
@@ -0,0 +1,116 @@
+    /// <param name="timeout">The timeout after which the <see cref="Task"/> should be faulted with a <see cref="TimeoutException"/> if it hasn't otherwise completed.</param>
+    /// <param name="cancellationToken">The <see cref="CancellationToken"/> to monitor for a cancellation request.</param>
+    /// <returns>The <see cref="Task"/> representing the asynchronous wait.</returns>
+    internal static async Task WaitAsync(Func<CancellationToken, Task> getTaskFunc, NpgsqlTimeout timeout,
```
It seems like the naming (and documentation) should express the fact that this method swallows unobserved exceptions, especially given the .NET 6.0 WaitAsync which has the same name but does not... Maybe WaitAsyncAndPreventUnobservedExceptions?

baal2000 replied:
If I understand correctly, a method's name that API accepts `Func<CancellationToken, Task> getTaskFunc` should start with "Execute". Then if I rename to `ExecuteWithTimeoutAndFakeCancellationWhilePreventingUnobservedExceptions` and update the comments, would that satisfy such a requirement?

roji replied:
I'm not 100% hung up on specific naming - just trying to make the code clearer. I think it would be better to separate out the concern of applying a timeout via the cancellation token (previously called ExecuteWithTimeout) and the concern of preventing unobserved exceptions. And yeah, Execute seems like better naming than Wait since .NET's WaitAsync accepts an already running Task, whereas Execute accepts a lambda which will be executed and return that Task.

ExecuteWithTimeoutAndFakeCancellationWhilePreventingUnobservedExceptions is quite a mouthful :) We could simply leave two separate methods (ExecuteWithTimeout, WaitAsyncWhilePreventingUnobservedExceptions) and call one over the other in the two call sites we have. Or if you feel strongly that there should be one method which takes care of both (while ideally calling a local ExecuteWithTimeout method), that's OK for me too.

baal2000 replied:
We should not call one over another because
- doing so in 2 places would be repeating yourself,
- splitting ExecuteWithTimeout from PreventingUnobservedExceptions would promote future misuse and more unobserved exceptions.

Considering that the functionality is private to NpgsqlConnector I'd vote for a shorter name like `ExecuteWithTimeoutAndCancellation` leaving the rest to the comment.

For a general-purpose API we'd add a parameter "preventUnobservedExceptions" that would self-document the API but we should not do that without a use case.

roji replied:
> splitting ExecuteWithTimeout from PreventingUnobservedExceptions would promote future misuse and more unobserved exceptions.

Technically I don't think that's true - ExecuteWithTimout simply implements timeout via the cancellation token, and by itself shouldn't create any misuse.

But I think we've spent too much time discussing this. Proper commenting should help make things clear.

baal2000 replied:
Respecting your opinion on how the code has to be structured: let me take another shot at this.

baal2000 replied:
The problem is that we need to "unobserve" the Task produced by `getTaskFunc`. That makes for an awkward `ExecuteWithTimeout` interface to satisfy the need to await on the composite Task and at the same time returning `getTaskFunc` Task for later optional chaining into `PreventingUnobservedExceptions`.

Feel free to make your own update if I am missing the right way to do this.

roji replied:
I'm OK with the code as it is... Let's get it merged.

roji commented on src/Npgsql/Internal/NpgsqlConnector.cs:
```diff
@@ -939,11 +939,20 @@ void Connect(NpgsqlTimeout timeout)
-            ? new EndPoint[] { new UnixDomainSocketEndPoint(socketPath) }
-            : (await GetHostAddressesAsync(timeout, cancellationToken))
-            .Select(a => new IPEndPoint(a, Port)).ToArray();
+                ? new EndPoint[] { new UnixDomainSocketEndPoint(socketPath) }
```
indentation

baal2000 replied:
This is how my IDE did it. Do you want me to add more idents or remove any?

```
        var endpoints = NpgsqlConnectionStringBuilder.IsUnixSocket(Host, Port, out var socketPath)
                                ? new EndPoint[] { new UnixDomainSocketEndPoint(socketPath) }
                                : Dns.GetHostAddresses(Host).Select(a => new IPEndPoint(a, Port)).ToArray();
```

v.s.
```
        var endpoints = NpgsqlConnectionStringBuilder.IsUnixSocket(Host, Port, out var socketPath)
        ? new EndPoint[] { new UnixDomainSocketEndPoint(socketPath) }
        : Dns.GetHostAddresses(Host).Select(a => new IPEndPoint(a, Port)).ToArray();
```

roji replied:
We simply do 4-space increases, so:

```c#
        var endpoints = NpgsqlConnectionStringBuilder.IsUnixSocket(Host, Port, out var socketPath)
            ? new EndPoint[] { new UnixDomainSocketEndPoint(socketPath) }
            : (await TaskTimeoutAndCancellation.WaitAsync(GetHostAddressesAsync, timeout, cancellationToken))
            .Select(a => new IPEndPoint(a, Port)).ToArray();
```

baal2000 replied:
Got it.

roji commented:
@baal2000 thanks for the changes (and the patience), here are some more comments.

baal2000 commented:
@roji 
Addressed the second round of the review comments.

roji reviewed (approved):
@baal2000 thanks again - LGTM, will just wait for the last bit moving the legacy code into the Shims directory and then merge.
